pennylane/measurements/classical_shadow.py

```diff
@@ -442,6 +442,131 @@ def process_state_with_shots(
+                created from the ``seed`` argument provided to ``.ClassicalShadowsMP``.
+
+        Returns:
+            tensor_like[int]: A tensor with shape ``(2, T, n)``, where the first row represents
+            the measured bits and the second represents the recipes used.
+        """
+        wire_map = {w: i for i, w in enumerate(wire_order)}
+        mapped_wires = [wire_map[w] for w in self.wires]
+        n_qubits = len(mapped_wires)
+        num_dev_qubits = len(state.shape) // 2
+
+        # seed the random measurement generation so that recipes
+        # are the same for different executions with the same seed
+        recipe_rng = np.random.RandomState(self.seed)
+        recipes = recipe_rng.randint(0, 3, size=(shots, n_qubits))
+
+        bit_rng = np.random.default_rng(rng)
+
+        obs_list = np.array(
+            [
+                [[0, 1], [1, 0]],  # X
+                [[0, -1j], [1j, 0]],  # Y
+                [[1, 0], [0, -1]],  # Z
+            ]
+        )
+
+        # the diagonalizing matrices corresponding to the Pauli observables above
+        diag_list = np.array(
+            [
+                [[1 / np.sqrt(2), 1 / np.sqrt(2)], [1 / np.sqrt(2), -1 / np.sqrt(2)]],
+                [[0.5 + 0.5j, 0.5 - 0.5j], [0.5 + 0.5j, -0.5 + 0.5j]],
+                [[1, 0], [0, 1]],
+            ]
+        )
+        obs = obs_list[recipes]
+        diagonalizers = diag_list[recipes]
+
+        # transpose the state so that the measured wires appear first
+        unmeasured_wires = [i for i in range(num_dev_qubits) if i not in mapped_wires]
+        transposed_state = np.transpose(
+            state,
+            axes=mapped_wires
+            + unmeasured_wires
+            + [w + num_dev_qubits for w in mapped_wires]
+            + [w + num_dev_qubits for w in unmeasured_wires],
+        )
+
+        outcomes = np.zeros((shots, n_qubits))
+        stacked_state = np.repeat(transposed_state[np.newaxis, ...], shots, axis=0)
+
+        for active_qubit in range(n_qubits):
+            # stacked_state loses a dimension each loop
+
+            # trace out every qubit except the first
+            num_remaining_qubits = num_dev_qubits - active_qubit
+            conj_state_first_qubit = ascii_letters[num_remaining_qubits]
+            stacked_dim = ascii_letters[num_remaining_qubits + 1]
+            remaining_dim = ascii_letters[:num_remaining_qubits]
+            traced_dim = remaining_dim[1:]
+
+            state_str = f"{stacked_dim}{remaining_dim}"
+            conj_state_str = f"{conj_state_first_qubit}{traced_dim}"
+            target_str = f"{stacked_dim}a{conj_state_first_qubit}"
+
+            first_qubit_state = np.einsum(
+                f"{state_str}{conj_state_str}->{target_str}",
+                stacked_state,
+            )
+
+            # sample the observables on the first qubit
+            probs = (np.einsum("abc,acb->a", first_qubit_state, obs[:, active_qubit]) + 1) / 2
```

Comment:
I was thinking: Currently we compute the probs in three different bases, then rotate the state, correct?
We could also just rotate first and the compute the probs in `Z` basis exclusively, reducing the second object in this contraction here from `(num_shots, 2, 2)` to `(2, 2)` :sweat_smile: 
But maybe that's an optimization for later :)

Reply:
Sounds a good idea! Will record this to the story for later optimization as well